### PR TITLE
Drop LSAN suppressions (almost)

### LINKED ISSUE #3071: Trigger garbage collector before exit to prevent valgrind's false positives
At least it worth to check whether this helps us with valgrind.

### LINKED ISSUE #8890: FP LeakSanitizer's leak reports when JIT is on
Found in 3.0.0-alpha1-16-g0e5a3cc21.

LeakSanitizer reports leaks on `sql-tap/uuid.test` run in `box_tuple_iterator`. There are few usages of this function in code and all looks correct from leaks POV. The issue seems to be related to JIT as it cannot be reproduced with JIT off. Otherwise although it is flaky it can be easily reproduced under load (`yes sql-tap/uuid.test | head -n 20 | xargs test/test-run.py --builddir ../build-asan-release -j16 --force`).

To get a repro you need ASAN-friendly allocators. It is WIP in scope of https://github.com/tarantool/tarantool/issues/7327. Temporarily available in https://github.com/nshy/tarantool/tree/small-asan.

```
==864465==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 9877 byte(s) in 83 object(s) allocated from:
    #0 0x564942c4ddc9 in malloc (/home/shiny/dev/tarantool/build-asan-release/src/tarantool+0x915dc9) (BuildId: 08a27b3b9efac51a4542f3e6792cb219dbaba87c)
    #1 0x564943d0005d in small_wrapper_alloc /home/shiny/dev/tarantool/src/lib/small/include/small/util.h:222:25
    #2 0x564943d0005d in mempool_alloc /home/shiny/dev/tarantool/src/lib/small/small/mempool_malloc.c:71:2
    #3 0x564943869034 in box_tuple_iterator /home/shiny/dev/tarantool/src/box/tuple.c:731:3
    #4 0x56496d5e5ebb  (<unknown module>)

SUMMARY: AddressSanitizer: 9877 byte(s) leaked in 83 allocation(s).
```

**Other similar issues**.

**Issue 2**
```
==722763==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 620 byte(s) in 10 object(s) allocated from:
    #0 0x55f13c58a299 in malloc (/home/shiny/dev/tarantool/build-asan-release/src/tarantool+0x8f0299) (BuildId: 8f67e24dd582992058f0172f0991094b1df87983)
    #1 0x55f13d63a5df in small_asan_alloc (/home/shiny/dev/tarantool/build-asan-release/src/tarantool+0x19a05df) (BuildId: 8f67e24dd582992058f0172f0991094b1df87983)
    #2 0x55f13d639bd8 in smalloc (/home/shiny/dev/tarantool/build-asan-release/src/tarantool+0x199fbd8) (BuildId: 8f67e24dd582992058f0172f0991094b1df87983)
    #3 0x55f13d1d5985 in runtime_tuple_new tuple.c
    #4 0x55f13d1d340f in box_tuple_update (/home/shiny/dev/tarantool/build-asan-release/src/tarantool+0x153940f) (BuildId: 8f67e24dd582992058f0172f0991094b1df87983)
    #5 0x55f166f51a96  (<unknown module>)

SUMMARY: AddressSanitizer: 620 byte(s) leaked in 10 allocation(s).
```
Reproducible as in the example above with the test ` box/update.test`.

**Issue 3**
```
=================================================================
==3350970==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 114968 byte(s) in 7 object(s) allocated from:
    #0 0x55aa60c8f3c9 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x11903c9) (BuildId: 840ceebdec4bf9f55a3e17a30442449c3073b4c9)
    #1 0x55aa62eaab35 in small_asan_alloc /home/shiny/dev/tarantool/src/lib/small/small/util.c:94:24
    #2 0x55aa62e9accd in slab_get /home/shiny/dev/tarantool/src/lib/small/small/slab_cache_asan.c:48:4
    #3 0x55aa62e9677e in ibuf_reserve_slow /home/shiny/dev/tarantool/src/lib/small/small/ibuf.c:93:23
    #4 0x55aa8c2e8ed4  (<unknown module>)

Direct leak of 3192 byte(s) in 14 object(s) allocated from:
    #0 0x55aa60c8f3c9 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x11903c9) (BuildId: 840ceebdec4bf9f55a3e17a30442449c3073b4c9)
    #1 0x55aa62eaab35 in small_asan_alloc /home/shiny/dev/tarantool/src/lib/small/small/util.c:94:24
    #2 0x55aa62e9c5d1 in mempool_alloc /home/shiny/dev/tarantool/src/lib/small/small/mempool_asan.c:67:4
    #3 0x55aa60dd8de0 in iterator* memtx_tree_index_create_iterator<true>(index*, iterator_type, char const*, unsigned int, char const*) /home/shiny/dev/tarantool/src/box/memtx_tree.cc:1616:3
    #4 0x55aa60d89578 in index_create_iterator_after(index*, iterator_type, char const*, unsigned int, char const*) /home/shiny/dev/tarantool/src/box/index.h:982:9
    #5 0x55aa60d884eb in box_index_iterator_after /home/shiny/dev/tarantool/src/box/index.cc:496:24
    #6 0x55aa8c2e18d5  (<unknown module>)

SUMMARY: AddressSanitizer: 118160 byte(s) leaked in 21 allocation(s).
```
for `engine/errinj_ddl.test.lua` in **debug** ASAN build (failure is not flaky).

**Issue 4**
```
=================================================================
==3355973==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 16424 byte(s) in 1 object(s) allocated from:
    #0 0x55a4127833c9 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x11903c9) (BuildId: 840ceebdec4bf9f55a3e17a30442449c3073b4c9)
    #1 0x55a41499eb35 in small_asan_alloc /home/shiny/dev/tarantool/src/lib/small/small/util.c:94:24
    #2 0x55a41498eccd in slab_get /home/shiny/dev/tarantool/src/lib/small/small/slab_cache_asan.c:48:4
    #3 0x55a41498a77e in ibuf_reserve_slow /home/shiny/dev/tarantool/src/lib/small/small/ibuf.c:93:23
    #4 0x55a3dac49239  (<unknown module>)

SUMMARY: AddressSanitizer: 16424 byte(s) leaked in 1 allocation(s).
```
for `box/reconfigure.test.lua` in **debug** ASAN build (failure is flaky).

### LINKED ISSUE #9213: ASAN: memory leak in luaT_push_key_def
Found in `3.0.0-alpha3-69-gba9b012ac`

Leak is reported with special allocators for ASAN (PR https://github.com/tarantool/tarantool/pull/8901) for `box/net.box_iproto_transactions_over_streams.test.lua`. I guess PR does not make a difference taking the backtrace into account. It is only reported in **debug** build. Looks like it is get suppressed in release mode. With `malloc_context_size=50` (default is `30`) in `ASAN_OPTIONS` issue is gone for debug mode too.

Build options:
```bash
CC=/usr/bin/clang CXX=/usr/bin/clang++ cmake -B build-asan-debug \
        -DCMAKE_BUILD_TYPE=Debug \
        -DENABLE_WERROR=ON \
        -DENABLE_ASAN=ON \
        -DENABLE_UB_SANITIZER=ON \
        -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=OFF \
        -DFIBER_STACK_SIZE=640Kb \
        -DENABLE_FUZZER=ON \
        -DTEST_BUILD=ON
```
```
=================================================================
==3320211==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1152 byte(s) in 6 object(s) allocated from:
    #0 0x55f70fa273c9 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x11903c9) (BuildId: 50b4f8d425d61717239211bb327e4c0a466dabf8)
    #1 0x55f71175f161 in key_def_dup /home/shiny/dev/tarantool/src/box/key_def.c:144:24
    #2 0x55f710b87458 in luaT_push_key_def /home/shiny/dev/tarantool/src/box/lua/key_def.c:71:29
    #3 0x55f710b17a32 in luaT_add_index_parts_methods(lua_State*, key_def const*) /home/shiny/dev/tarantool/src/box/lua/space.cc:323:2
    #4 0x55f710b15354 in lbox_fillspace(lua_State*, space*, int) /home/shiny/dev/tarantool/src/box/lua/space.cc:523:3
    #5 0x55f710b123b1 in box_lua_space_new(lua_State*, space*) /home/shiny/dev/tarantool/src/box/lua/space.cc:634:2
    #6 0x55f710b11f76 in box_lua_space_new_or_delete(trigger*, void*) /home/shiny/dev/tarantool/src/box/lua/space.cc:664:3
    #7 0x55f710e222b6 in trigger_run_list(rlist*, void*) /home/shiny/dev/tarantool/src/lib/core/trigger.cc:119:7
    #8 0x55f710e20836 in trigger_run /home/shiny/dev/tarantool/src/lib/core/trigger.cc:152:9
    #9 0x55f7101a785e in space_cache_replace /home/shiny/dev/tarantool/src/box/space_cache.c:187:6
    #10 0x55f71028c12c in alter_space_do(txn_stmt*, alter_space*) /home/shiny/dev/tarantool/src/box/alter.cc:1066:2
    #11 0x55f71025419b in on_replace_dd_space(trigger*, void*) /home/shiny/dev/tarantool/src/box/alter.cc:2493:4
    #12 0x55f710e222b6 in trigger_run_list(rlist*, void*) /home/shiny/dev/tarantool/src/lib/core/trigger.cc:119:7
    #13 0x55f710e20836 in trigger_run /home/shiny/dev/tarantool/src/lib/core/trigger.cc:152:9
    #14 0x55f710187804 in space_on_replace /home/shiny/dev/tarantool/src/box/space.c:881:11
    #15 0x55f7103213a1 in txn_commit_stmt /home/shiny/dev/tarantool/src/box/txn.c:644:13
    #16 0x55f71037b665 in box_process_rw /home/shiny/dev/tarantool/src/box/box.cc:474:6
    #17 0x55f7103a21c6 in box_process1 /home/shiny/dev/tarantool/src/box/box.cc:3608:9
    #18 0x55f7103a9f3d in box_update /home/shiny/dev/tarantool/src/box/box.cc:3850:9
    #19 0x55f710577398 in sql_constraint_create /home/shiny/dev/tarantool/src/box/sql.c:1550:6
    #20 0x55f710578e44 in sql_check_create /home/shiny/dev/tarantool/src/box/sql.c:1664:9
    #21 0x55f710957a96 in sqlVdbeExec /home/shiny/dev/tarantool/src/box/sql/vdbe.c:2189:6
    #22 0x55f7109a042c in sqlStep /home/shiny/dev/tarantool/src/box/sql/vdbeapi.c:133:8
    #23 0x55f71099f39c in sql_step /home/shiny/dev/tarantool/src/box/sql/vdbeapi.c:155:9
    #24 0x55f7105879b0 in sql_execute /home/shiny/dev/tarantool/src/box/execute.c:218:8
    #25 0x55f7105874e6 in sql_prepare_and_execute /home/shiny/dev/tarantool/src/box/execute.c:281:6
    #26 0x55f710588760 in box_process_sql /home/shiny/dev/tarantool/src/box/execute.c:313:11
    #27 0x55f70fab8a36 in tx_process_sql(cmsg*) /home/shiny/dev/tarantool/src/box/iproto.cc:2440:6
    #28 0x55f710d45e75 in cmsg_deliver /home/shiny/dev/tarantool/src/lib/core/cbus.c:373:2
    #29 0x55f710d54df4 in fiber_pool_f /home/shiny/dev/tarantool/src/lib/core/fiber_pool.c:64:3

Direct leak of 384 byte(s) in 2 object(s) allocated from:
    #0 0x55f70fa273c9 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x11903c9) (BuildId: 50b4f8d425d61717239211bb327e4c0a466dabf8)
    #1 0x55f71175f161 in key_def_dup /home/shiny/dev/tarantool/src/box/key_def.c:144:24
    #2 0x55f710b87458 in luaT_push_key_def /home/shiny/dev/tarantool/src/box/lua/key_def.c:71:29
    #3 0x55f710b17a32 in luaT_add_index_parts_methods(lua_State*, key_def const*) /home/shiny/dev/tarantool/src/box/lua/space.cc:323:2
    #4 0x55f710b15354 in lbox_fillspace(lua_State*, space*, int) /home/shiny/dev/tarantool/src/box/lua/space.cc:523:3
    #5 0x55f710b123b1 in box_lua_space_new(lua_State*, space*) /home/shiny/dev/tarantool/src/box/lua/space.cc:634:2
    #6 0x55f710b11f76 in box_lua_space_new_or_delete(trigger*, void*) /home/shiny/dev/tarantool/src/box/lua/space.cc:664:3
    #7 0x55f710e222b6 in trigger_run_list(rlist*, void*) /home/shiny/dev/tarantool/src/lib/core/trigger.cc:119:7
    #8 0x55f710e20836 in trigger_run /home/shiny/dev/tarantool/src/lib/core/trigger.cc:152:9
    #9 0x55f7101a785e in space_cache_replace /home/shiny/dev/tarantool/src/box/space_cache.c:187:6
    #10 0x55f71028c12c in alter_space_do(txn_stmt*, alter_space*) /home/shiny/dev/tarantool/src/box/alter.cc:1066:2
    #11 0x55f71025ce6d in on_replace_dd_truncate(trigger*, void*) /home/shiny/dev/tarantool/src/box/alter.cc:2906:3
    #12 0x55f710e222b6 in trigger_run_list(rlist*, void*) /home/shiny/dev/tarantool/src/lib/core/trigger.cc:119:7
    #13 0x55f710e20836 in trigger_run /home/shiny/dev/tarantool/src/lib/core/trigger.cc:152:9
    #14 0x55f710187804 in space_on_replace /home/shiny/dev/tarantool/src/box/space.c:881:11
    #15 0x55f7103213a1 in txn_commit_stmt /home/shiny/dev/tarantool/src/box/txn.c:644:13
    #16 0x55f71037b665 in box_process_rw /home/shiny/dev/tarantool/src/box/box.cc:474:6
    #17 0x55f7103a21c6 in box_process1 /home/shiny/dev/tarantool/src/box/box.cc:3608:9
    #18 0x55f7103aa7bd in box_upsert /home/shiny/dev/tarantool/src/box/box.cc:3870:9
    #19 0x55f7103ab4d6 in space_truncate(space*) /home/shiny/dev/tarantool/src/box/box.cc:3901:6
    #20 0x55f7103aa950 in box_truncate /home/shiny/dev/tarantool/src/box/box.cc:3911:3
    #21 0x55f71098155c in sqlVdbeExec /home/shiny/dev/tarantool/src/box/sql/vdbe.c:3828:7
    #22 0x55f7109a042c in sqlStep /home/shiny/dev/tarantool/src/box/sql/vdbeapi.c:133:8
    #23 0x55f71099f39c in sql_step /home/shiny/dev/tarantool/src/box/sql/vdbeapi.c:155:9
    #24 0x55f7105879b0 in sql_execute /home/shiny/dev/tarantool/src/box/execute.c:218:8
    #25 0x55f7105874e6 in sql_prepare_and_execute /home/shiny/dev/tarantool/src/box/execute.c:281:6
    #26 0x55f710588760 in box_process_sql /home/shiny/dev/tarantool/src/box/execute.c:313:11
    #27 0x55f70fab8a36 in tx_process_sql(cmsg*) /home/shiny/dev/tarantool/src/box/iproto.cc:2440:6
    #28 0x55f710d45e75 in cmsg_deliver /home/shiny/dev/tarantool/src/lib/core/cbus.c:373:2
    #29 0x55f710d54df4 in fiber_pool_f /home/shiny/dev/tarantool/src/lib/core/fiber_pool.c:64:3

SUMMARY: AddressSanitizer: 1536 byte(s) leaked in 8 allocation(s).

```

### LINKED ISSUE #10055: Leak detected by asan in `box/session_settings` test.
```
[008] box/session_settings.test.lua                                   [ pass ]
[008] 
[008] 
[008] [Instance "box" returns with non-zero exit code: 1]
[008] 
[008] [test-run server "box"] Last 15 lines of the log file /tmp/t/008_box/box.log:
[008]     #7 0x55731c3d6385 in box_iterator_next /__w/tarantool/tarantool/src/box/index.cc:588:6
[008]     #8 0x557346ce9331  (<unknown module>)
[008] 
[008] Direct leak of 132 byte(s) in 2 object(s) allocated from:
[008]     #0 0x55731c34707e in malloc (/__w/tarantool/tarantool/src/tarantool+0x8d107e) (BuildId: 7c422d7e3542336ec6f35363b2a1524a6cfcc4c3)
[008]     #1 0x55731d3d58f5 in small_asan_alloc /__w/tarantool/tarantool/src/lib/small/small/util.c:94:24
[008]     #2 0x55731d3d4ab5 in smalloc /__w/tarantool/tarantool/src/lib/small/small/small_asan.c:57:5
[008]     #3 0x55731cf389a5 in runtime_tuple_new /__w/tarantool/tarantool/src/box/tuple.c:138:27
[008]     #4 0x55731cf37634 in tuple_new /__w/tarantool/tarantool/src/box/tuple.h:801:9
[008]     #5 0x55731cf37634 in box_tuple_new /__w/tarantool/tarantool/src/box/tuple.c:845:22
[008]     #6 0x55731c4c87b3 in session_settings_index_get /__w/tarantool/tarantool/src/box/session_settings.c:261:12
[008]     #7 0x55731c3d1d48 in box_index_get /__w/tarantool/tarantool/src/box/index.cc:394:11
[008]     #8 0x557346ce0e05  (<unknown module>)
[008] 
[008] SUMMARY: AddressSanitizer: 5478 byte(s) leaked in 81 allocation(s).
```
Workflow run: https://github.com/tarantool/tarantool/actions/runs/9224121473/job/25386134593?pr=10017#step:8:9142

### LINKED ISSUE #10211: get rid of LSAN suppressions
Some of suppressions are quite wide so we do not know clearly our leaks.

```
leak:fiber_cxx_invoke
leak:Exception::operator new
leak:libc.so*
leak:lj_BC_FUNCC
```

### LINKED ISSUE #10495: swim crash on Tarantool shutdown
It is hard to reproduce on current master (3.3.0-entrypoint). Much eaiser to reproduce with PR #10207 (without `"swim: fix gc after Tarantool shutdown started"` commit of course). 

Run `test/test-run.py --force --builddir ../build-dev app-luatest/shutdown_test.lua -p test_swim_shutdown`. The Tarantool server crashes with backtrace:
```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6,
    no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007a3d15ca5463 in __pthread_kill_internal (threadid=<optimized out>, signo=6)
    at pthread_kill.c:78
#2  0x00007a3d15c4c120 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007a3d15c334c3 in __GI_abort () at abort.c:79
#4  0x00005f0d1eb9ed5d in crash_signal_cb (signo=11, siginfo=0x5f0d2aa6cfb0,
    context=0x5f0d2aa6ce80) at /home/shiny/dev/tarantool/src/lib/core/crash.c:204
#5  <signal handler called>
#6  0x00005f0d1eba68a0 in fiber_is_dead (f=0x0)
    at /home/shiny/dev/tarantool/src/lib/core/fiber.h:1176
#7  0x00005f0d1eba8aa6 in fiber_cancel (f=0x0)
    at /home/shiny/dev/tarantool/src/lib/core/fiber.c:630
#8  0x00005f0d1ebf5f86 in swim_gc (swim=0x5f0d2aae9590)
    at /home/shiny/dev/tarantool/src/lib/swim/swim.c:2257
#9  0x00005f0d1eb91abe in lua_swim_gc (L=0x40ed1378)
    at /home/shiny/dev/tarantool/src/lua/swim.c:166
#10 0x00005f0d1ec07ac7 in lj_BC_FUNCC () at buildvm_x86.dasc:811
#11 0x00005f0d1ec22f16 in gc_call_finalizer (g=0x40ed13b8, L=0x40ed1378,
    mo=0x7ffffce9d2a0, o=0x40c49070)
    at /home/shiny/dev/tarantool/third_party/luajit/src/lj_gc.c:514
#12 0x00005f0d1ec23b96 in lj_gc_finalize_cdata (L=0x40ed1378)
    at /home/shiny/dev/tarantool/third_party/luajit/src/lj_gc.c:593
#13 0x00005f0d1ec2d7b3 in cpfinalize (L=0x40ed1378, dummy=0x0, ud=0x0)
    at /home/shiny/dev/tarantool/third_party/luajit/src/lj_state.c:275
#14 0x00005f0d1ec07eb8 in lj_vm_cpcall () at buildvm_x86.dasc:1204
#15 0x00005f0d1ec2d921 in lua_close (L=0x40ed1378)
    at /home/shiny/dev/tarantool/third_party/luajit/src/lj_state.c:308
#16 0x00005f0d1eb6d43d in tarantool_lua_free ()
    at /home/shiny/dev/tarantool/src/lua/init.c:1252
#17 0x00005f0d1e8d502e in tarantool_free ()
    at /home/shiny/dev/tarantool/src/main.cc:562
#18 0x00005f0d1e8d603b in main (argc=1, argv=0x5f0d2aa63e78)
    at /home/shiny/dev/tarantool/src/main.cc:1104
```

It is easier to reproduce with #10207 because we start to call `lua_close` on Tarantool exit there. So all Lua objects are collected.
Yet the issue is possible in `3.2` as after swim worker fiber is finished Lua can collect swim object just due to memory pressure.

Introduced in `3.2.0-entrypoint-128-g1f41645219`.

### PR DESCRIPTION
**Pull request by nshy**
Factored out from all-in-one PR https://github.com/tarantool/tarantool/pull/10207

Closes #10211
Closes #3071
Closes #8890
Closes #10055
Closes #9213
Closes #10495

Related EE PR https://github.com/tarantool/tarantool-ee/pull/835

### CODE REVIEW

File: src/main.cc [hunk 1]
```diff
@@ -558,6 +558,8 @@ tarantool_free(void)
 	coio_shutdown();
 
 	box_lua_free();
+	/* Lua may have reference to engine tuples. */
+	tarantool_lua_free();
```
Comment (igormunkin) [inline src/main.cc:562]: Finally.

**Review by p7nov — APPROVED** [review]

File: src/box/lua/space.cc [hunk 2]
```diff
@@ -772,12 +772,16 @@ box_lua_space_delete(struct lua_State *L, uint32_t id)
...
+	(void)trigger;
+	struct lua_State *L = tarantool_L;
 	struct space *space = (struct space *) event;
 
 	if (space_by_id(space->def->id) != NULL) {
 		box_lua_space_new(L, space);
 	} else {
+		/* tarantool_L is freed on Tarantool shutdown. */
+		if (L == NULL)
+			return 0;
```
Comment (locker) [inline src/box/lua/space.cc:784]: We should delete the trigger in `box_lua_space_free` instead.

File: asan/lsan.supp [hunk 3]
```diff
@@ -4,77 +4,9 @@
...
-# TODO(gh-9213): remove when the issue is fixed
-leak:luaT_push_key_def
-
-# TODO(gh-8890): remove when the issue is fixed
-leak:box_tuple_update
-leak:box_tuple_iterator
-leak:box_index_iterator_after
-leak:ibuf_reserve_slow
-leak:box_check_slice
+# FIXME suppress until gh-10184 is fixed
```
Comment (locker) [inline asan/lsan.supp:10]: Why not fix it now?

File: asan/lsan.supp [hunk 4]
```diff
@@ -4,77 +4,9 @@
 # File format:
 #leak:*
 
-# test: app/crypto.test.lua
-# source: /usr/lib/x86_64-linux-gnu/libcrypto.so
-leak:CRYPTO_zalloc
+leak:sql_prepare_and_execute
+leak:sqlParser
```
Comment (locker) [inline asan/lsan.supp:8]: Are there any plans to fix these leaks as well?

File: src/box/lua/space.cc (hunk 2 above)
Comment (nshy) [reply src/box/lua/space.cc:784]: Fixed.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/lua/init.c b/src/box/lua/init.c
index 6bdc27a186..5d607b4071 100644
--- a/src/box/lua/init.c
+++ b/src/box/lua/init.c
@@ -995,4 +995,5 @@ void
 box_lua_free(void)
 {
 	box_lua_iproto_free();
+	box_lua_space_free();
 }
diff --git a/src/box/lua/space.cc b/src/box/lua/space.cc
index ce9830f206..79276fadad 100644
--- a/src/box/lua/space.cc
+++ b/src/box/lua/space.cc
@@ -772,16 +772,12 @@ box_lua_space_delete(struct lua_State *L, uint32_t id)
 static int
 box_lua_space_new_or_delete(struct trigger *trigger, void *event)
 {
-	(void)trigger;
-	struct lua_State *L = tarantool_L;
+	struct lua_State *L = (struct lua_State *) trigger->data;
 	struct space *space = (struct space *) event;
 
 	if (space_by_id(space->def->id) != NULL) {
 		box_lua_space_new(L, space);
 	} else {
-		/* tarantool_L is freed on Tarantool shutdown. */
-		if (L == NULL)
-			return 0;
 		box_lua_space_delete(L, space->def->id);
 	}
 	return 0;
@@ -921,7 +917,7 @@ void
 box_lua_space_init(struct lua_State *L)
 {
 	/* Register the trigger that will push space data to Lua. */
-	assert(tarantool_L == L);
+	on_alter_space_in_lua.data = L;
 	trigger_add(&on_alter_space, &on_alter_space_in_lua);
 
 	lua_getfield(L, LUA_GLOBALSINDEX, "box");
@@ -1029,3 +1025,9 @@ box_lua_space_init(struct lua_State *L)
 	luaL_setfuncs(L, space_internal_lib, 0);
 	lua_pop(L, 1);
 }
+
+void
+box_lua_space_free(void)
+{
+	trigger_clear(&on_alter_space_in_lua);
+}
diff --git a/src/box/lua/space.h b/src/box/lua/space.h
index 1155e8fb39..7421332de0 100644
--- a/src/box/lua/space.h
+++ b/src/box/lua/space.h
@@ -41,6 +41,10 @@ struct lua_State;
 void
 box_lua_space_init(struct lua_State *L);
 
+/** Free module resources. */
+void
+box_lua_space_free(void);
+
 #if defined(__cplusplus)
 } /* extern "C" */
 #endif /* defined(__cplusplus) */

```

</details>

File: asan/lsan.supp (hunk 4 above)
Comment (nshy) [reply asan/lsan.supp:8]: Added an issue #10532:

```diff
diff --git a/asan/lsan.supp b/asan/lsan.supp
index 17e690abf3..e18e6bbe60 100644
--- a/asan/lsan.supp
+++ b/asan/lsan.supp
@@ -4,9 +4,10 @@
 # File format:
 #leak:*

+# to be removed in scope of gh-10532
 leak:sql_prepare_and_execute
 leak:sqlParser

-# FIXME suppress until gh-10184 is fixed
+# to be removed in scope of gh-10184
 leak:lua_yaml_encode
 leak:l_load
```

File: src/box/lua/space.h [hunk 5]
```diff
@@ -41,6 +41,10 @@ struct lua_State;
 void
 box_lua_space_init(struct lua_State *L);
 
+/** Free module resources. */
```
Comment (locker) [inline src/box/lua/space.h:44]: Let's not add a comment - for consistency with `box_lua_space_init()`, which doesn't have one. Checkpatch doesn't require comments for init/free.

**Review by locker — APPROVED** [review]

File: src/box/lua/space.h (hunk 5 above)
Comment (nshy) [reply src/box/lua/space.h:44]: Fixed.

**Comment by locker** [comment]
Cherry-picked commit `swim: fix crash on Tarantool shutdown` to 3.2.

**Pull request merged**